go/systests/teams_test.go

```diff
@@ -2105,3 +2127,517 @@ func TestTeamHiddenGenerationRotateRace(t *testing.T) {
 	alice.addTeamMember(parentName.String(), bob.username, keybase1.TeamRole_ADMIN)
 	bob.waitForMetadataUpdateGregor("added back")
 }
+
+func mustAppend(t *testing.T, a keybase1.TeamName, b string) keybase1.TeamName {
+	ret, err := a.Append(b)
+	require.NoError(t, err)
+	return ret
+}
+
+func mustCreateSubteam(t *testing.T, tc *libkb.TestContext,
+	name keybase1.TeamName) keybase1.TeamID {
+	parent, err := name.Parent()
+	require.NoError(t, err)
+	id, err := teams.CreateSubteam(context.TODO(), tc.G, name.LastPart().String(),
+		parent, keybase1.TeamRole_NONE)
+	require.NoError(t, err)
+	return *id
+}
+
+func loadTeamTree(t *testing.T, tmctx libkb.MetaContext, notifications *teamNotifyHandler,
+	teamID keybase1.TeamID, username string, converter teams.ITreeloaderStateConverter,
+	teamFailures []keybase1.TeamName) ([]keybase1.TeamTreeMembership, error) {
+	var err error
+	if converter == nil {
+		err = teams.LoadTeamTreeMemberships(tmctx, teamID, username)
+	} else {
+		err = teams.LoadTeamTreeMembershipsWithConverter(tmctx, teamID, username, converter)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	var results []keybase1.TeamTreeMembership
+	var expectedCount *int
+	got := 0
+loop:
+	for {
+		select {
+		case res := <-notifications.TeamTreeMembershipsDoneCh:
+			// We don't immediately break in this case because we are not guaranteed to receive this
+			// notification last by the RPC layer. So we wait until all of the messages are
+			// received. Even if there were errors, we should still get exactly this many
+			// notifications in TeamTreeMembershipsPartialCh.
+			expectedCount = &res
+		case res := <-notifications.TeamTreeMembershipsPartialCh:
+			got++
+			results = append(results, res)
+			s, err := res.Result.S()
+			require.NoError(t, err, "should never happen")
+			switch s {
+			case keybase1.TeamTreeMembershipStatus_OK:
+			case keybase1.TeamTreeMembershipStatus_ERROR:
+				require.Contains(t, teamFailures, res.TeamName,
+					"unexpectedly got an error while loading team: %s", res.Result.Error().Message)
+			}
+		case <-time.After(10 * time.Second):
+			t.Fatalf("timed out waiting for team tree notifications")
+		}
+		if expectedCount != nil && *expectedCount == got {
+			break loop
+		}
+	}
+	return results, nil
+}
+
+func checkTeamTreeResults(t *testing.T, expected map[string]keybase1.TeamRole,
+	failureTeamNames []keybase1.TeamName, results []keybase1.TeamTreeMembership) {
+	require.Equal(t, len(expected)+len(failureTeamNames),
+		len(results), "got right number of results back")
+	for _, result := range results {
```

Comment:
this does not test for duplicate results. I.e. I could receive the same notification n times and the test would pass. you could make a map to catch duplicates....

Reply:
Does the length check on 2195 cover that case?